src/dataflow/src/render/upsert.rs

```diff
@@ -398,18 +405,91 @@ where
+                                            )))
+                                        } else {
+                                            actual_value.and_then(|non_err_val| {
+                                                let full_row_count = non_err_val.iter().count();
+                                                if full_row_count != permutation.arity() {
+                                                    Err(DataflowError::DecodeError(
+                                                        DecodeError::Text(format!(
+                                                            "Failed to ingest data, \
+                                                            found value with {} columns, \
+                                                            expected {}",
+                                                            full_row_count,
+                                                            permutation.key_arity()
+                                                        )),
+                                                    ))
+                                                } else {
+                                                    row_packer.clear();
+
+                                                    let borrow =
+                                                        datum_vec.borrow_with(&non_err_val);
+
+                                                    for keep_index in thinning {
+                                                        row_packer.push(borrow[*keep_index]);
+                                                    }
+
+                                                    let r = row_packer.finish_and_reuse();
+                                                    Ok(r)
+                                                }
+                                            })
+                                        }
+                                    });
+
                                     let old_value = if let Some(new_value) = &new_value {
-                                        current_values.insert(decoded_key, new_value.clone())
+                                        current_values
+                                            .insert(decoded_key.clone(), new_value.clone())
                                     } else {
                                         current_values.remove(&decoded_key)
                                     };
+
+                                    // This closure re-uses `row_packer` to re-assemble rows with
+                                    // help from the `Permutation`
+                                    let mut repack_value = |row: Row| {
+                                        let mut borrow =
+                                            datum_vec.borrow_with_many(&[&decoded_key, &row]);
+                                        permutation.permute_in_place(&mut borrow);
+
+                                        row_packer.extend(borrow.iter());
```

Comment:
Please check that permut_in_place truncates the output so we have the expected number of column here. I'm on mobile :)

Reply:
`permute_in_place` makes sure that the vector is the same length as the permutation, which was created with thinning of the key + the row, so it should be the correct length, always